### PR TITLE
TestBot: Implement inject_mocks method

### PR DESCRIPTION
**Pull request by nvzard**
Method to inject property in the plugin
object and discard the existing assignments.

### CODE REVIEW

File: errbot/backends/test.py [hunk 1]
```diff
@@ -461,6 +461,26 @@ def assertCommandFound(self, command, timeout=5):
...
+        }
+        testbot.inject_mocks(hello, mock_dict)
+        assert 'blah' in exec_command('!hello')
+        """
+        plugin = self.bot.plugin_manager.get_plugin_obj_by_name(PluginName)
+
+        if plugin is None:
+            raise Exception(f'"{PluginName}" is not loaded.')
+        for field, mock_obj in MockDict.items():
+            if hasattr(plugin, field):
```
Comment (gbin) [inline errbot/backends/test.py]: fail early:
```python
if not hasattr(plugin, field):
     raise ValueError(f'No property/attribute named "{field}" attached.')

setattr(plugin, field, mock_obj)
```

Comment (nvzard) [reply errbot/backends/test.py]: @gbin the testbot is not able to load the DummyPlugin and that is the reason tests are failing. Am I missing something in the logic here? I tried both ways to fetch the loaded plugin:
```
self.bot.plugin_manager.get_plugin_obj_by_name(PluginName)
```
as well as:
```
self.bot.plugin_manager.plugins[PluginName]
```

File: docs/user_guide/plugin_development/testing.rst [hunk 2]
```diff
@@ -121,6 +121,39 @@ What we need now is get access to the instance of our plugin itself. Fortunately
...
+
+    class MyPlugin(BotPlugin):
+        @botcmd
+        def mycommand(self, message, args):
+            return self.mycommand_helper()
+
+        def mycommand_helper(self):
+            return (requests.get(URL).status_code)
+
+What we need now is to somehow replace the method making the request with out mock object and `inject_mocks` method comes in handy.
```
Comment (meetmangukiya) [inline docs/user_guide/plugin_development/testing.rst]: our mock object*

File: docs/user_guide/plugin_development/testing.rst [hunk 3]
```diff
@@ -121,6 +121,39 @@ What we need now is get access to the instance of our plugin itself. Fortunately
...
+
+Refer https://pypi.org/project/mock/ for more information about mock.
```
Comment (meetmangukiya) [inline docs/user_guide/plugin_development/testing.rst]: Errbot is just python3, python3 ships with mock in `unittest.mock`

File: docs/user_guide/plugin_development/testing.rst [hunk 4]
```diff
@@ -121,6 +121,39 @@ What we need now is get access to the instance of our plugin itself. Fortunately
...
+
+.. code-block:: python
+
+    from mock import MagicMock
```
Comment (meetmangukiya) [inline docs/user_guide/plugin_development/testing.rst]: `unittest.mock`

File: docs/user_guide/plugin_development/testing.rst [hunk 5]
```diff
@@ -121,6 +121,39 @@ What we need now is get access to the instance of our plugin itself. Fortunately
...
+
+    extra_plugin_dir = '.'
+
+    def test_mycommand_helper(testbot):
+        helper_mock = MagicMock(return_value='200')
+        mock_dict = {'mycommand_helper', helper_mock}
```
Comment (meetmangukiya) [inline docs/user_guide/plugin_development/testing.rst]: `mock_dict = {'mycommand_helper': helper_mock}`, mind the colon.

File: errbot/backends/test.py [hunk 6]
```diff
@@ -461,6 +461,25 @@ def assertCommandFound(self, command, timeout=5):
         self.bot.push_message(command)
         assert 'not found' not in self.bot.pop_message(timeout)
 
+    def inject_mocks(self, PluginName: str, MockDict: dict):
```
Comment (meetmangukiya) [inline errbot/backends/test.py]: class like name for arguments :confused: ?

**Pull request merged**